### PR TITLE
Dp3 206 - Error messages on registration/log-in screen should be more diagnostic

### PR DESCRIPTION
**Pull request by JccE**
This fixes the form and function of login and registration error validations. e2e tests included.

### CODE REVIEW

File: src/pages/login/login.scss [hunk 1]
```diff
@@ -33,4 +35,14 @@ page-login {
   #login-container {
   }
 
+  .qspacer {
+    font-size: 0;
+    height: 250px;
+    line-height: 0;
+    border: 0px;
+    /* testing only */
```
Comment (BinSquare) [inline src/pages/login/login.scss:43]: reminder to eliminate test comments

File: src/pages/login/login.scss [hunk 2]
```diff
@@ -5,6 +5,8 @@ page-login {
     flex-direction: column;
     justify-content: flex-start;
     align-items: center;
+    height: 100vh;
```
Comment (BinSquare) [inline src/pages/login/login.scss:8]: careful about using vh, is this tested on mobile sizes?

File: src/components/login/login.ts [hunk 3]
```diff
@@ -1,49 +1,87 @@
...
+  selector: "login",
+  templateUrl: "login.html",
 })
-
 export class LoginComponent {
-
-	user: any = {};
-	private errors: any = [];
+  user: any = {};
+  counter: any = 0;
```
Comment (BinSquare) [inline src/components/login/login.ts:13]: counter seems to be only numbers from what I can see, is there a reason to use any specifically?

File: src/components/register/register.html [hunk 4]
```diff
@@ -1,69 +1,66 @@
...
-      </div>
-      <div *ngIf="registerSelected "style="display: flex; justify-content: flex-end">
-          <h6 id="register-link"> back to Login </h6>
-      </div>
-
-     </div>
-
-     <!-- <div class="qspacer"></div> -->
+    </div>
+    <div style="display: flex; justify-content: flex-end">
```
Comment (BinSquare) [inline src/components/register/register.html:58]: no action needed, question: 

Is there any possibility the style specifications can get moved to the css files?

File: src/components/register/register.html [hunk 5]
```diff
@@ -1,69 +1,66 @@
...
           </div>
+          <input type="password" name="passwd" [(ngModel)]="user.passwd2" id="pw2">
+        </div>
+        <div id="button-container">
+          <button ion-button color="buttonBlue" (click)="submitRegistration()">Register</button>
+        </div>
+        <div id="password-rules">
+          <p *ngIf="checkError('invalid_passwd')" style="color:red">* passwords must be at least 10 characters in
+            length, with at least one digit, special character, and uppercase letter. </p>
+          <p *ngIf="!checkError('invalid_passwd')"> * passwords must be at least 10 characters in length, with at least
```
Comment (BinSquare) [inline src/components/register/register.html:53]: Think angular lets you switch styles based on condition here but it's nitpicky - don't need to spend more time on this unless you choose to

File: cypress/integration/login-error-validation-test.js [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+  const emailBlankMessage = "enter an email";
+  const passwordInvalidMessage = "enter valid password";
+  const passwordBlankMessage = "enter a password";
+
+  beforeEach(() => {
+    cy.visit(baseUrl);
+    cy.viewport(viewportWidth, viewportHeight);
+  });
+
+  it("displays proper validation erros if submitted form is blank ", () => {
```
[the remark this answers is not in the document]
Comment (JccE) [reply cypress/integration/login-error-validation-test.js:29]: Done

File: cypress/integration/login-error-validation-test.js [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+    cy.get(emailBlankId).should(($error) => {
+      expect($error).to.contain.text(emailBlankMessage);
+    });
+    cy.get(passwordBlankId).should(($error) => {
+      expect($error).to.contain.text(passwordBlankMessage);
+    });
+  });
+
+  it("displays proper validation errors if submitted form is partial ", () => {
+    cy.get(nameInput).type(emailInvalid);
```
[the remark this answers is not in the document]
Comment (JccE) [reply cypress/integration/login-error-validation-test.js:40]: Done

File: cypress/integration/registration-error-validation-test.js [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+  const passwordBlankMessage = "enter a password";
+  const passwordNoMatchMessage = "passwords don't match";
+
+  beforeEach(() => {
+    cy.viewport(viewportWidth, viewportHeight);
+    cy.visit(baseUrl);
+    cy.get(registerLink).click();
+  });
+
+  it("displays proper validation erros if submitted form is blank ", () => {
```
[the remark this answers is not in the document]
Comment (JccE) [reply cypress/integration/registration-error-validation-test.js:36]: Fixed

File: cypress/integration/login-error-validation-test.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (JccE) [reply cypress/integration/login-error-validation-test.js:29]: Fixed

File: cypress/integration/login-error-validation-test.js [hunk 9]
```diff
@@ -0,0 +1,51 @@
+
+describe("Error validations on Login page should be visible ", function () {
+  const baseUrl = Cypress.env("baseUrl");
+  const viewportWidth = Cypress.env("viewportWidth");
+  const viewportHeight = Cypress.env("viewportHeight");
+
+  const emailInvalid = "j#j.com";
```
[the remark this answers is not in the document]
Comment (JccE) [reply cypress/integration/login-error-validation-test.js:7]: This is noted as a todo item. I added a list of invalid emails to start with.

File: src/pages/login/login.scss (hunk 2 above)
Comment (JccE) [reply src/pages/login/login.scss:8]: For the most part, I have not touched the css or tested the mobile version. Only desktop Web at the moment

File: src/components/login/login.ts (hunk 3 above)
Comment (JccE) [reply src/components/login/login.ts]: Done

File: src/pages/login/login.scss (hunk 2 above)
Comment (BinSquare) [reply src/pages/login/login.scss:8]: The mobile versions are built with the same code base - you can test for the mobile sizes using device toolbar in your browser. You've added a height of 100vh here, is it an accident or is it needed?

File: src/components/register/register.html (hunk 5 above)
Comment (JccE) [reply src/components/register/register.html:53]: Tried to fix, I'll have to come back to this

File: src/pages/login/login.scss (hunk 2 above)
Comment (JccE) [reply src/pages/login/login.scss:8]: I did not add it...

Comment (BinSquare) [reply src/pages/login/login.scss:8]: I see what happened - nonetheless, these changes that are non-related to your feature is being introduced in your PR. Seems like you branched off from a starting point that included Dom's changes here: https://github.com/EktaFlow/docent-base/pull/178/commits/3b3bd2e90b3bb4462992f3c28f7bacad030ae1b8

I'd recommend parsing through your commits and cherry picking only the commits related to your PR.

File: cypress/integration/login-error-validation-test.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (BinSquare) [reply cypress/integration/login-error-validation-test.js:7]: Gotcha, one quick way to finish this off would be to iterate through the list of invalid inputs for the test. Then you'd have a fairly decent confidence that at least the regex for the emails are working as designed. 

I am going to insist that we knock this out in this PR as just checking "j#j.com" is not representative enough to say that we're catching most invalid emails. Let me know if you need help on this

[the remark this answers is not in the document]
Comment (JccE) [reply cypress/integration/login-error-validation-test.js:7]: Complete.

File: cypress/integration/registration-error-validation-test.js [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+    "email@example",
+    "email@-example.com",
+    "email@example.web",
+    "email@111.222.333.44444",
+    "email@example..com",
+    "Abc..123@example.com"
+  ]
+
+
+  const emailValid = "j@k.com";
```
Comment (BinSquare) [inline cypress/integration/registration-error-validation-test.js:38]: same thing as the invalid emails, "j@k.com" doesn't represent good coverage of valid emails.

There's some here: https://gist.github.com/cjaoude/fd9910626629b53c4d25
And some here: https://www.w3resource.com/javascript/form/email-validation.php

File: cypress/integration/login-error-validation-test.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (BinSquare) [reply cypress/integration/login-error-validation-test.js:7]: The test for login-error-validation-test.js hasn't been changed yet, think you may have forgotten to port it over?

File: cypress/support/commands.js [hunk 11]
```diff
@@ -19,6 +19,35 @@ Cypress.Commands.add("login", (email, password) => {
 });
 //
 
+
```
Comment (BinSquare) [inline cypress/support/commands.js:22]: Is this code still necessary to be left in the codebase?

Comment (JccE) [reply cypress/support/commands.js]: Removed old comments and code

File: cypress/integration/registration-error-validation-test.js (hunk 10 above)
Comment (JccE) [reply cypress/integration/registration-error-validation-test.js]: Provided a new list of valid emails!

File: cypress/integration/login-error-validation-test.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (JccE) [reply cypress/integration/login-error-validation-test.js]: Ported over!

**Comment by BinSquare** [comment]
Thank you for your work, some aspects of this have been reworked and merged into master.

**Pull request closed without merging**